docs: Add a guide on how to document REST API endpoints.

eeshangarg opened the pull request:
@timabbott, @YagoGG: FYI :)

timabbott commented:
@YagoGG @brickbite I'd love it if you two can give it a review.  @YagoGG as someone who's worked on this system, and @brickbite as someone who hasn't :)

timabbott commented:
(For @brickbite, you can build it as ReadTheDocs would using `tools/build-docs` and access it at 127.0.0.1:9991/docs)

YagoGG commented on docs/tutorials/documenting-api-endpoints.md:
```diff
@@ -0,0 +1,121 @@
+file in `templates/zerver/help/rest-endpoints.md`, so that it can be displayed
+in the sidebar on the `/api` page.
+
+## Usage examples.
```
I think it's better if we remove the period here, as we don't use it for titles along the docs.

brickbite replied:
Agree on removing the period for better consistency

eeshangarg replied:
This is a typo, so it wasn't intended lol! :)

YagoGG commented on docs/tutorials/documenting-api-endpoints.md:
```diff
@@ -0,0 +1,121 @@
+## Usage examples.
+
+We display usage examples in three languages: Python, JavaScript and `curl`.
+For JavaScript and `curl` we simply recommend copying and pasting the examples
```
It may be worth mentioning that the examples reflect the behavior of our JS bindings (maybe a link to the binding's repo?).

brickbite commented:
@YagoGG got it up, that was right. thanks!

@eeshangarg 
Might be good to include a link to show where contributions will end up (https://zulipchat.com/api). I hadn't gone through it before, and had to poke around to find it

brickbite commented on docs/tutorials/documenting-api-endpoints.md:
```diff
@@ -0,0 +1,121 @@
+1. Make the desired API call inside the function. If our Python bindings don't
+   have a dedicated method for a specific API call, you may either use
+   `client.call_endpoint` or add a dedicated function to the
+   [`zulip` package](https://github.com/zulip/python-zulip-api/tree/master/zulip).
```
Not sure if this part is formatted as intended.

**Screenshot (both zulip and package are clickable):**
![image](https://user-images.githubusercontent.com/21148168/40938139-c444d200-67f5-11e8-8a83-2b1a223b7b6c.png)

brickbite commented:
On the content overall: the process of going about this seems clear, with mentions of specific files to make edits to and where to get examples. I can also get an idea of how things are connected by reading it, so that's good!

eeshangarg commented:
@YagoGG, @brickbite: Thank you so much for your feedback! I addressed all of your points! Please let me know if there is anything else you'd like me to change! :)

@timabbott: FYI :)

timabbott commented:
Great, merged, after adding a few more changes in ce4f00bd8402359708bf1dcdbf1cb3cc95274d7e (the main thing being that I think the right flow is for them to write the test with a print statement, and get the fixture that way).  A few things that this makes me notice we could improve:
* We could get rid of the step to add things to the `test_` functions, which feels janky (and instead just call all the functions in the `TEST_FUNCTIONS` dict)
* We could eventually get rid of the `TEST_FUNCTIONS` dict step too, by just making these methods on a test class.
* If we wanted to really get fancy, I could imagine a flag that makes the `test_against_fixture` function auto-output the JSON fixtures for you if the fixture doesn't exist and/or even add it to the output, replacing the manual-feeling copy/paste flow.  May be more work than it's worth, though.
* It'd be super cool if we had a test that would compare the `REQ` arguments against the tests.  I think it might be possible to do this sort of check by some convoluted path involving (1) patching `has_request_variables` to store `post_params` as an attribute on the view function (2) having a mapping to find the view function, ideally just doing a lookup using `urls.py` so it's automated and (3) comparing the list of `post_params` to the documented arguments list.

@YagoGG FYI for this idea.
